### ISSUE TITLE
`ptr_arg` should trigger for `Option<&PathBuf>`

### ISSUE THREAD

**Issue by BenjaminBrienen** [rust-lang/rust-clippy#15504]
### Summary

A parameter of type `Option<&PathBuf>` could instead of of type `Option<&Path>`.

### Lint Name

ptr_arg

### Reproducer

I tried this code:

```rust
fn foo(maybe_path: Option<&PathBuf>) {
    if let Some(path) = maybe_path {
        File::create(path).unwrap();
    }
}
```

I expected to see this happen:

ptr_arg triggers

Instead, this happened:

It doesn't

### Version

```text
rustc 1.89.0 (29483883e 2025-08-04)
binary: rustc
commit-hash: 29483883eed69d5fb4db01964cdf2af4d86e9cb2
commit-date: 2025-08-04
host: x86_64-unknown-linux-gnu
release: 1.89.0
LLVM version: 20.1.7
```

**Comment by ada4a** [comment]
This doesn't seem to have anything to do with the `Option` -- if I remove it, I still get no lint:
```rs
fn foo(path: &PathBuf) {
    File::create(path).unwrap();
}
```

It seems like `File::open` is the actual culprit here: the type of its argument is `impl AsRef<Path>`, and I think the lint is not smart enough to realize that `&Path` could be used for that just as well? Because if I change it to a method on `PathBuf` implemented through `Deref<Target=Path>`, the lint fires again:
```rs
fn foo(path: &PathBuf) {
    let _ = path.as_os_str();
}
```

But you're right as well: if we go full circle, even that doesn't work when we reintroduce the `Option`:
```rs
fn foo(maybe_path: Option<&PathBuf>) {
    if let Some(path) = maybe_path { 
        let _ = path.as_os_str();
    }
}
```

So I guess this issue should really be split in two.

**Comment by BenjaminBrienen** [comment]
Oh, I used that method as a contrived example. I replaced PathBuf with Path in my code and it worked.

**Comment by ada4a** [comment]
> it worked

You mean the code compiled after the change? If so, then of course, the change _is_ valid, it's just that the lint doesn't suggest it.. Or maybe I'm misunderstanding you?

**Comment by BenjaminBrienen** [comment]
Looks like the "real world example" is the same issue with `AsRef<Path>`. See test-utils/src/lib.rs `try_ensure_file_contents`. It doesn't trigger a lint if you change it to `file: &PathBuf`.

`file` is only used as `file: impl AsRef<Path>` and `&*file` in that method.

**Comment by ada4a** [comment]
You mean [this](https://github.com/rust-lang/rust/blob/99ba5565678a51c2488322a5e75d5b59e323b498/src/tools/rust-analyzer/crates/test-utils/src/lib.rs#L484)? That was a bit of a search, please just give a link next time 😅 

But yeah, it's most likely the `std::fs::read_to_string` [here](https://github.com/rust-lang/rust/blob/99ba5565678a51c2488322a5e75d5b59e323b498/src/tools/rust-analyzer/crates/test-utils/src/lib.rs#L485) that trips up the lint

**Comment by BenjaminBrienen** [comment]
This gives no lints:

```rs
pub fn foo(bar: &PathBuf, bazzy: Option<&PathBuf>) {
    let _ = bar;
    let _ = bazzy;
}
```

This only gives a lint for the first argument:

```rs
pub fn foo(bar: &PathBuf, bazzy: Option<&PathBuf>) {
    let z = bar;
    let zz = bazzy;
}
```

I expect that both arguments trigger the lint.

**Comment by zihan0822** [comment]
Not really sure whether we should introduce this. We did not lint `Option<&Vec<_>>`, `Option<&String>` etc as well, not just `Option<&PathBuf>`.

**Comment by ada4a** [comment]
Fwiw I think those would be good suggestions as well -- but getting the additional logic right will definitely be a challenge

**Comment by ada4a** [comment]
@BenjaminBrienen the underscore thing is intentional -- it was added in https://github.com/rust-lang/rust-clippy/pull/15105 to avoid churn during rapid prototyping

EDIT: though in that case I'm not sure why this code (from an earlier example) does lint:
```rs
fn foo(path: &PathBuf) {
    let _ = path.as_os_str();
}
```

EDIT 2: I think I know now: in the lint, we look, for each parameter, for the first place in the body of the function that kind of concreticizes the type of the parameter. In the case like this:
```rs
fn foo(p: &PathBuf) {
    let _ = p;
}
```
it's, unsurprisingly, the let stmt. And in the case of those, we check if the binding's name starts with a `_`, in which case we don't bother linting, to avoid churn.

But in the example before, that first-use place is actually the method call `path.as_os_str()`, and so we don't do the `_` check, and so still lint. I'd say that's pretty reasonable, since we _are_ using the binding in some way
